keps/sig-node/3759-sidecar-containers/README.md

```diff
@@ -0,0 +1,1243 @@
+  sidecar is healthy. Restart of sidecar containers is a best effort
+- allow to enforce security boundaries to sidecar containers different that
+  other containers. For example, allow Istio to run privileged to configure ip
+  tables and disable other containers from doing so
+- (alpha) support containers termination ordering
+
+## Proposal
+
+The proposal is to introduce a `restartPolicy` field to Init containers and use
+it to indicate that an initContainer is a sidecar container. The only supported
+value proposed in this KEP is `Always`. Kubelet will start these containers in
+the order with other Init containers, but instead of waiting for its completion,
+it will wait for the container startup completion (startup probe succeeded or if
+no startup probe defined, `postStart` handler completed).
+
+Other values for `restartPolicy` of containers will not be accepted and
+containers will follow the logic is currently implemented (documented
+[here](https://kubernetes.io/docs/concepts/workloads/pods/init-containers/#:~:text=if%20a%20pod's%20init%20container%20fails%2C%20the%20kubelet%20repeatedly%20restarts%20that%20init%20container%20until%20it%20succeeds.%20however%2C%20if%20the%20pod%20has%20a%20restartpolicy%20of%20never%2C%20and%20an%20init%20container%20fails%20during%20startup%20of%20that%20pod%2C%20kubernetes%20treats%20the%20overall%20pod%20as%20failed.)
+and more details can be found in the next section "Future use of `restartPolicy`
+field").
+
+Sidecar containers will not block Pod completion - if all regular containers
+complete, sidecar containers will be terminated.
+
+During the sidecar startup stage it will behave similarly to Init containers. If
+Pod `restartPolicy` is `Never`, sidecar container failed during startup will NOT
+be restarted and the whole Pod will fail. If Pod `restartPolicy` is `Always` or
+`OnFailure`, it will be restarted.
+
+Once sidecar container is started (postStart completed and startup probe
+succeeded), this containers will be restarted even when the Pod `restartPolicy`
+is `Never` or `OnFailure`. Furthermore, sidecar containers will be restarted
+even during Pod termination.
+
+In order to minimize OOM kills of sidecar containers, the OOM adjustment for
+these containers will match or exceed the OOM score adjustment of regular
+containers in the Pod. This intent to solve the issue
+https://github.com/kubernetes/kubernetes/issues/111356 
+
+As part of this KEP we also will be enabling for sidecar containers:
+- PreStart and PostStop lifecycle handlers for sidecar containers
+- All probes (startup, readiness, liveness)
+- Readiness probes of sidecars will be used to determine the whole Pod
+  readiness, but only after the main containers initialization started.
+  Readiness probes will not start running before the Pod initialization stage
+  completed 
+
+Since readiness probes will not start for sidecar containers till after
```

Comment:
This goes to my above comment - it says readiness probes don't start but doesn't explain WHY that decision is correct.

Reply:
Mostly to make sure that all readiness probes will start at the same time. Also, there is a strange notion of sidecar becoming not ready while other Init containers are running. Will this indicate that sidecar needs to restart? Or that initialization stage keeps running?

Splitting startup and readiness eliminates these questions

Reply:
I don't know if I buy it.  Readiness is there for a container to indicate that it is consumable.  Moving ahead without waiting for that feels wrong.

> Also, there is a strange notion of sidecar becoming not ready while other Init containers are running. Will this indicate that sidecar needs to restart?

Readiness does not cause restarts.

> Splitting startup and readiness eliminates these questions

You are, in effect saying "if you care about making sure your sidecar is ready, you have to use a `startupProbe`".  Does that feel weird to you?

Reply:
I'm getting in my head thru the scenario when sidecar wants to indicate that necessary initialization complete and other initialization can be started, but we are not ready to mark the pod as ready as we are doing some final initialization steps. Maybe even pinging from the sidecar other containers to be ready.

@howardjohn wdyt from the scenarios perspective?

Reply:
Pod readiness is only barely relevant here - a container can be ready without the pod being ready.  This is a discussion of a single container.

Reply:
My expectation was that once startup passes, the other containers could initialize and readiness probes would start (So readiness probes are the same - always immediately after startup passes).

But I hadn't considered the "no startup probe case"; given these are brand new usages of Pod we could say its required in order to have any of the startup blocking order.

From my POV, I don't care too much what probe signals it, so whatever is most consistent for K8s seems reasonable

Reply:
yes, I think it will be easier to just follow the existing pattern with startup and readiness probes. I now remember where this came from. From Risks and Mitigations section:

> Another potential problem may be that sidecars will wait for some condition to
mark itself “started” that cannot be met with the new pattern. For example, wait
for other containers to fully start. As sidecar will not succeed startup probes,
other Init containers will not run and Pod will be stuck on initialization. For
example, Knative's sidecar does aggressive probing of the user's container to
ensure they're ready prior to marking the sidecar ready itself (not for startup
probes though). This prevents the Pod from being included K8s ready endpoints.
However this problem will be discovered right away during the testing stage and
changes will need to be made to sidecar implementation.

Reply:
this is related to existing sidecars that just blindly ported to the new model. I am not sure how important this scenario is, but I also don't see any issues with splitting probes.

Reply:
@SergeyKanzhelev what do you mean with splitting probes exactly?

I don't see why the knative scenario wouldn't work if their sidecar reports as ready when it is ready. Then, when the main container is also ready, the pod will be added to the endpoints. 

Therefore, I don't see why sidecars won't wait on _readiness_ to continue.

Am I missing something?

Reply:
imagine you have a sidecar that looks at other containers readiness before marking itself ready. If kubelet will wait for ready signal (instead of "started"), than the container will never get ready - no containers will be started before sidecar initialization complete.

Splitting probes means that once startup probe succeeded, other Init containers initialization will happen and eventually regular containers will be started. Readiness probe for sidecar will start at later stage and kubelet will only look for ready signal to mark Pod as ready and do things with Endpoints. But not for the lifecycle control.

Reply:
> imagine you have a sidecar that looks at other containers readiness before marking itself ready.

This is challenging.  There *are* use-cases out there where one container serves readiness for another container.  For example, suppose your main app's real mechanism for detecting readiness is only exposed on  localhost, and you don't control it (e.g. an image you got from a vendor).  You can run a sidecar which does receives kubelet's probes, converts those to localhost, and serves the results.  In that case I would still configure it as the readiness probe OF THE "real" container, but to a port which just happens to be served by the helper.  The helper may have its own readiness.

So is there really a reason for one container's readiness to depend on a different container's readiness?  This feels like a gross violation of the principle of readiness.  @evankanderson - can you speak to this?  Does knative really do this?

Reply:
I'm not defending the pattern and received this as a point against readiness during one of the meetings.

If this pattern exists - @thockin what issues do you see with making "started" decision separate from "ready"? Readiness probes semantically are not about lifecycle, more about connectivity. Startup and liveness are about lifecycle and looks like a better fit here.

Reply:
I don't think it's the worst thing to depend on starupProbe for this, but it's "non-obvious", so I would like to be sure that a) this is really what we want; and b) that we document it clearly for this use-case.  For example, startupProbs DO NOT block subsequent "plain" containers from starting.  We know `initContainers` are different - let's be clear about how.

Reply:
Knative implements a "pass-through" of the original pod readiness to implement sub-second readiness probing (I believe a KEP was raised, but there were concerns about kubelet load in the general case).

Particularly for small, quick-starting containers like go binaries, we found increasing the tempo of readiness checks provided a 30%-ish reduction in cold start times.

It's disgusting, but we replace the container's declared probes with a probe of the sidecar, and pass the original container's probes to the sidecar as an argument of some type (I forget which type).

Reply:
More generally, one might imagine that (even without the dependency on another pod), a sidecar might want to start up, indicate that further processing can begin, but continue loading state and indicating that the Pod as a whole is not _ready_ until the state has finished loading.

(For example, this might include Istio starting up, installing the correct firewall rules in the Pod, and then marking `startup done` while it loads the mesh state. If subsequent init containers attempt to access the network, they will be blocked until the state is loaded, but if the init containers do _something else_, then they can proceed in parallel.)

Reply:
> `startupProbe`s DO NOT block subsequent "plain" containers from starting

Is there a notion of "subsequent 'plain' containers"? I'd always viewed the ordering of initContainers as a special thing.

Reply:
> Particularly for small, quick-starting containers like go binaries, we found increasing the tempo of initial readiness checks provided a 30%-ish reduction in cold start times.

This is, in part, why we added startupProbes.

> It's disgusting

Your words, not mine :)

> but we replace the container's declared probes with a probe of the sidecar

I can interpret this 2 ways, which one is right:

1) The readinessProbe field of the user-defined container is changed to point to a port on the sidecar.
2) The readinessProbe of the sidecar container acts as a proxy to the user-defined container's port. 

> a sidecar might want to start up, indicate that further processing can begin, but continue loading state and indicating that the Pod as a whole is not ready until the state has finished loading

This is awkward - I'm ready enough to use but not really ready enough to USE use.  I get the Istio case.  It's not even that unbelievable, it's just ... clunky.  I'd put the traffic capture into a synchronous init container, then the sidecar into a non-blocking one, and "ready" means "can accept connections" not "have all state needed".

That said, how hypopthetical is that case?  Maybe  @howardjohn has an opinion?

Reply:
> Is there a notion of "subsequent 'plain' containers"? I'd always viewed the ordering of initContainers as a special thing.

We don't guarantee sequential start, but it is sequential and I believe in Hyrum's law...

Reply:
> > `startupProbe`s DO NOT block subsequent "plain" containers from starting
> 
> Is there a notion of "subsequent 'plain' containers"? I'd always viewed the ordering of initContainers as a special thing.

I think this is referring to the fact that regular containers are not blocked on startupProbe execution. `postStart` lifecycle handlers are blocking and executed sequentially for regular containers while the startup probes for regular containers are all running in parallel. So there is a sequencing for regular containers, it just doesn't depend on startup probes.

Thinking about it - maybe just waiting for `postStart` to complete can be enough for sidecars. This will make containers behave similarly and will not require the probes loop to start very fast so startup will not be slowed down by it. Wdyt?

Reply:
The Istio case (in https://github.com/kubernetes/enhancements/pull/3761#discussion_r1089485312) is, at least for Istio, not something we would ever do. That is already the current state (just split between init container and app container) and its broken because we don't want to expose user's containers to a broken network.

Reply:
I feel that making startup to only wait for `postStart` solves all the scenarios and in fact easier to implement. Any onbjections on this thread? I am rewriting this commit...

Reply:
`postStart` blocking for readiness has some debugging challenges in how logs are reported (they aren't) and the status is displayed (I forget the details, but it doesn't signal "the pod is waiting", it looks pretty obscure IIRC). Perhaps we improve that situation, but it doesn't feel "right" to me. That being said, it meets our use cases so I am fine with it if others are

Reply:
@evankanderson I didn't get an answer to this question:

>> but we replace the container's declared probes with a probe of the sidecar
>
> I can interpret this 2 ways, which one is right:
> 
> 1) The readinessProbe field of the user-defined container is changed to point to a port on the sidecar.
> 2) The readinessProbe of the sidecar container acts as a proxy to the user-defined container's port.

Can you help me understand?

@SergeyKanzhelev my preference order for this design point is:

1) Use readiness probes (and maybe ensure that the container is "ready" after startupProbe succeeds - no need to wait for both)
2) Use startup probes for this and document it well
3) use postStart for this and make that more visible - tell me that my pod is "starting" in the API

What do you think?

Reply:
This was discussed at today's Sidecar WG meeting. The concern of not being able to express "not ready, but started" was dominant. @smarterclayton mentioned other prior KEPs trying to take dependency on Ready state that wasn't accepted.

There was agreement on using the Startup probes to indicate this status. @matthyx offered to craft the proposal on [marking Pods as started](https://kubernetes.slack.com/archives/C0BP8PW9G/p1675194507712149) in general and applying it to sidecars once implemented.

As with every decision we keep the door open to change something after Alpha's feedback.

Reply:
Sorry, I've been at a conference.

> > but we replace the container's declared probes with a probe of the sidecar
>
> I can interpret this 2 ways, which one is right:

> 1. The readinessProbe field of the user-defined container is changed to point to a port on the sidecar.
>1. The readinessProbe of the sidecar container acts as a proxy to the user-defined container's port.

It's actually (3):

The `readinessProbe` on the user container is _removed_. The `queue-proxy` binary has a `readinessProbe` defined, whose status depends on the results of the `queue-proxy` health-checking the user's container.

This allows us to implement a "retry quickly until container is started, then reduce pace to configured frequency" policy to improve initial readiness time without increasing steady-state load.